### DISCUSSION TITLE
How setup a cross compilation pipeline with android + win32 + iOS?

### DISCUSSION THREAD

**Discussion by mamcx** [ipetkov/crane#463] [General]
I have a fairly simple project that is unfortunately hard to compile for all my targets. I'm trying to port all to nix + crane so I eliminate all my custom scripts that broke all the time (and maybe cut CI times).

My current strategy ([as talked here](https://github.com/ipetkov/crane/discussions)) is trying to duplicate what https://github.com/dpc/fedimint/blob/master/flake.nix has, but that is pure guesswork in what to do. The other is to make this from scratch, but then I have the major complication that built with NDK broke so much and as of today I don't know how to fix it in a way that plays nice with current versions (I have a solution for an old NDK version that only work in a very OLD CI machine).

In short: The MAJOR step is how to make NDK compile with Rustqlite as dep as "simply" as `nix build -L .#droid`. I probably need this is 2 discussions but the first step is the initial setup...

My current structure is:

```bash
~/Proyectos/Apps took 5s
❯ tree
.
├── IronAge -> /Users/mamcx/Proyectos/IronAge
├── RustStore
│     ├── Cargo.lock
│     ├── Cargo.toml
│     ├── client
│     │     └── ui
│     │         ├── Cargo.toml
│     │         └── src
│     │             └── lib.rs
│     ├── corelib
│     │     ├── Cargo.toml
│     │     └── src
│     │         └── lib.rs
│     ├── corelogic
│     │     ├── Cargo.toml
│     │     └── src
│     │         └── lib.rs
│     ├── server
│     │     ├── Cargo.toml
│     │     └── src
│     │         └── main.rs
│     ├── sync
│     │     ├── Cargo.toml
│     │     └── src
│     │         └── main.rs
│     └── syncdb
│         ├── Cargo.toml
│         └── src
│             └── main.rs
├── flake.nix
```

This is a project made with only the deps listed in `Cargo.tolm` with not source yet, to simplify it.

I want/need:

- Run the CI set of test: `test, clippy`
- Compile for Windows
- Compile for Android
- Compile for iOS (not critical now)
- Run on macOS M1
- Run on circleCI on top of a nix machine

So, how to make this work "from scratch"?

**Comment by dpc** [comment]
> My current strategy ([as talked here](https://github.com/ipetkov/crane/discussions)) is trying to duplicate what https://github.com/dpc/fedimint/blob/master/flake.nix has

Just so you're aware we've extracted all of that stuff into https://github.com/rustshop/flakebox with it's own tutorials etc.

Ping me on Flakebox's Matrix channel and I can try to help (though sick ATM, so less available). But I must warn you that cross-compilation is just hard and often comes down to grinding looking for combinations of things that work. I am responsible for both Fedimint and an propriety application on top that cross-compiles to bunch of Android archs, macos, ios, so I know it can be done.

**Reply by mamcx** [reply]
Ah great (also sick, so tomorrow I will ping)

**Reply by mamcx** [reply]
I put a repo with the naked setup with deps at https://github.com/mamcx/RustStoreNix

**Reply by mamcx** [reply]
Ok, this hit a dead-end. May simplify the question to just how to build for android? I think is the most complex of all.
